### PR TITLE
Fix scope and injector issues in local compilation mode

### PR DESCRIPTION
**Pull request by pmvald**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/empty_fields.local.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
 export class FooModule {}
 FooModule.ɵfac = …;
 FooModule.ɵmod = …;
-FooModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({});
+FooModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({imports: [[]]});
```
Comment (crisbeto) [inline packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/empty_fields.local.js:4]: nit: could we avoid pushing the extra empty array for some of the easier-to-detect cases? E.g. `ts.isArrayLiteralExpression(node) && node.elements.length === 0`.

File: packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/imports_exports.local.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 export class AModule {}
 AModule.ɵfac = function AModule_Factory(t) { return new (t || AModule)(); };
 AModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AModule });
-AModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [A1Component, A2Component] });
+AModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[A1Component, A2Component]] });
```
Comment (crisbeto) [inline packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/imports_exports.local.js:4]: Why do we have the extra array here?

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 3]
```diff
@@ -468,10 +468,7 @@ export class ComponentDecoratorHandler implements
           viewProviders: wrappedViewProviders,
           i18nUseExternalIds: this.i18nUseExternalIds,
           relativeContextFilePath,
-          rawImports: (rawImports !== null && ts.isArrayLiteralExpression(rawImports) &&
-                       rawImports.elements.length > 0) ?
-              new WrappedNodeExpr(rawImports) :
-              undefined,
+          rawImports: (rawImports !== null) ? new WrappedNodeExpr(rawImports) : undefined,
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:471]: nit: you don't need the parentheses around the `rawImports !== null`.

File: packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/imports_exports.local.js (hunk 2 above)
Comment (devversion) [reply packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/ng_modules/imports_exports.local.js:4]: I think we can address this + your comment above by simply special-casing if it's an actual array literal (and keep logic like before).

File: packages/core/src/render3/deps_tracker/api.ts [hunk 4]
```diff
@@ -29,6 +29,14 @@ interface ScopeData {
   isPoisoned?: boolean;
 }
 
+/**
+ * Represents scope data for standalone components as calculated during runtime by the deps
+ * tracker.
+ */
+interface StandaloneCompScopeData extends ScopeData {
+  ngModules: Set<NgModuleType<any>>;
```
Comment (devversion) [inline packages/core/src/render3/deps_tracker/api.ts:37]: Nit: add a comment on why standalone components track `ngModules`

**Review by devversion — APPROVED** [review]
lgtm conceptually. just some comments + from Kristiyan

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts (hunk 3 above)
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts]: Done!

File: packages/core/src/render3/deps_tracker/api.ts (hunk 4 above)
Comment (pmvald) [reply packages/core/src/render3/deps_tracker/api.ts:39]: Done!

**Review by devversion — APPROVED** [review]
LGTM. Not sure if it's necessary to address the nit (new presubmit etc)-no need to request re-review from my end

File: packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts [hunk 5]
```diff
@@ -445,12 +445,22 @@ export class NgModuleDecoratorHandler implements
...
 
-      if (rawExports !== null && ts.isArrayLiteralExpression(rawExports) && rawExports.elements) {
-        injectorMetadata.imports.push(...rawExports.elements.map(n => new WrappedNodeExpr(n)));
+        if (ts.isArrayLiteralExpression(exp)) {
+          // If array expression, then add it entry-by-entry to the injector imports
+          if (exp.elements) {
+            injectorMetadata.imports.push(...exp.elements.map(n => new WrappedNodeExpr(n)));
+          }
+        } else {
+          // if not array expression, then add it as it is to the injector's imports field.
```
[the remark this answers is not in the document]
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/ng_module/src/handler.ts]: Done!

**Review by crisbeto — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit b44533ba39334514960a7a28bc96d3843edc8e15.

**Pull request closed without merging**